### PR TITLE
[PP] microbatch split config

### PR DESCRIPTION
**Pull request by H-Huang**
Allow the input batch to be split on the sequence dimension in pipeline parallelism (remove the requirement for batch_size >= num stages)

depends on https://github.com/pytorch/pytorch/pull/148458

### CODE REVIEW

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -359,6 +359,14 @@ def __init__(self):
                 If using looped schedules, this still specifies the number of physical ranks, not the number
                 of stages.  Stages per rank are inferred from split points degree, and schedule.""",
         )
+        self.parser.add_argument(
+            "--experimental.pipeline_parallel_batch_split_dim",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py:363]: It seems PP has been in `experimental` for a while. Do you think it's time we extract `pipeline_parallel` into a standalone section and put all configs over there?
It doesn't have to happen in this PR.

File: torchtitan/distributed/pipeline.py [hunk 2]
```diff
@@ -119,17 +121,18 @@ def build_pipeline_schedule(
             f"of stages ({num_total_stages}) which may result in a bubble in the pipeline."
         )
 
-    # validate that the batch size is divisible by the number of microbatches otherwise we'll hang or error during training
```
Comment (tianyu-l) [inline torchtitan/distributed/pipeline.py:122]: I have several questions here:
1. If `pipeline_parallel_batch_split_dim == 0`, what would happen if `if job_config.training.batch_size % num_total_stages != 0`?
2. If `pipeline_parallel_batch_split_dim` is on the sequence dim or other dims, don't we need similar checks in the extremal cases e.g. seq_len < num_stages
3. Btw this divisibility requirement seems not exactly the same as "batch_size >= num stages" you mentioned in the PR summary.
